packages/amplify_storage_s3/example/integration_test/main_test.dart

```diff
@@ -0,0 +1,121 @@
+    late String? lastUploadedKey;
+
+    // Returns a text file to use for testing, writing if it does not exist.
+    Future<File> getTemporaryFile() async {
+      final path = '${(await getTemporaryDirectory()).path}/$exampleFileName';
+      final file = File(path);
+      if (!(await file.exists())) {
+        await file.writeAsString('Upload me to s3 to see if I work.');
+      }
+      return file;
+    }
+
+    Future<int> getCountFromListFiles() async {
+      final result = await Amplify.Storage.list(options: listOptions);
+      return result.items.length;
+    }
+
+    Future<void> deleteAllGuestFiles() async {
+      final result = await Amplify.Storage.list(options: listOptions);
+      for (StorageItem item in result.items) {
+        await Amplify.Storage.remove(key: item.key);
+      }
+    }
+
+    setUpAll(() async {
+      Amplify.addPlugins([AmplifyAuthCognito(), AmplifyStorageS3()]);
+      await Amplify.configure(amplifyconfig);
+
+      await deleteAllGuestFiles();
+    });
+
+    tearDownAll(() async {
+      await deleteAllGuestFiles();
+    });
+
+    testWidgets('should upload a file', (WidgetTester tester) async {
+      final file = await getTemporaryFile();
+      final initialCount = await getCountFromListFiles();
+
+      final key = '$exampleFileName${new DateTime.now().toString()}';
+      Map<String, String> metadata = <String, String>{};
+      metadata['name'] = exampleFileName;
+      metadata['desc'] = 'A test file';
+      S3UploadFileOptions options = S3UploadFileOptions(
+          accessLevel: StorageAccessLevel.guest, metadata: metadata);
+      final result = await Amplify.Storage.uploadFile(
+          key: key, local: file, options: options);
+      lastUploadedKey = result.key;
+      expect(lastUploadedKey, key);
+
+      final finalCount = await getCountFromListFiles();
+      expect(initialCount + 1, finalCount);
+    });
+
+    testWidgets('should list files and get expected attributes',
+        (WidgetTester tester) async {
+      if (lastUploadedKey == null) {
+        fail('No uploaded file to verify.');
+      }
+      final result = await Amplify.Storage.list(options: listOptions);
+      expect(result.items.length, greaterThan(0));
+      final uploadedStorageItem =
+          result.items.firstWhere((element) => element.key == lastUploadedKey);
+      expect(uploadedStorageItem.key, lastUploadedKey);
+      expect(uploadedStorageItem.lastModified?.day,
+          new DateTime.now().day); // was uploaded today
+      expect(uploadedStorageItem.eTag?.isNotEmpty, true);
+      expect(uploadedStorageItem.size, greaterThan(0));
+    });
+
+    testWidgets('should get the URL of an uploaded file',
+        (WidgetTester tester) async {
+      if (lastUploadedKey == null) {
+        fail('No uploaded file to verify.');
+      }
+      final result = await Amplify.Storage.getUrl(key: lastUploadedKey!);
+      expect(result.url.isNotEmpty, true);
```

Comment:
Maybe this is a bit nitpicky, but can this ever be empty? It seems like this test might just be checking to make sure that getUrl succeeds and does not throw, in which case there are some methods to express exactly that, which may help with how the test reads.

```dart
await expectLater(Amplify.Storage.getUrl(key: lastUploadedKey!), completes);
```

or, less wordy:
```dart
final getUrl = Amplify.Storage.getUrl(key: lastUploadedKey!);
await expectLater(getUrl, completes);
```

Reply:
Yeah I don't think assertion is quite right. I wanted to do **some** validation on the URL result here, though emptiness probably not good. Initially I tried to validate the key appears in the URL, but then I found it gets URI escaped only on iOS (seems like an issue?) in this case, bc key has special characters. So, assertion was not super straightforward. I guess I could try:
a) using a key with no special characters or otherwise asserting with URI escaping
b) validate another way, thought about looking for 's3'  or just asserting it's a valid URL
c) do what you suggested and just assert completion

I'm kinda leaning towards b now. If that's too cumbersome, might just do c like you suggested.

Reply:
Gotcha, I understand now. We could go all the way and perform a GET request with the returned url and compare the contents, if you want.

Reply:
updated assertion a la option b